spec/p2p/v0.34/README.md

```diff
@@ -0,0 +1,140 @@
+
+Depending on whether the node is a seed node or or not, the PEX reactor constantly runs either `crawlPeersRoutine` or `ensurePeers()` respectively. 
+If the node is a seed node, `crawlPeersRoutine` reads peer information randomly from the address book, tries to dial the peer and requests from them information on other peers.
+
+When a node receives information about other peers from a seed note, it sends another request to the node asking for more peers. 
+
+A node also learns of other peers when they try to connect to it. 
+
+### Dialing peers
+
+Every node connects to a set of peers to whom it sends requests via the Switch. 
+
+The PEX reactor ensures a node is connected to peers by running a routine to keep dialing peers in the background until a threshold of connected peers is reached (`MaxNumOutboundPeers`). This way, if a peers' address has been added via an incoming request, it will eventually be dialed. Peers to dial are chosen with some configurable bias for unvetted peers. The bias should be lower when we have fewer peers and can increase as we obtain more, ensuring that our first peers are more trustworthy, but always giving us the chance to discover new good peers. 
+
+As the number of outgoing peers is limited, the reactor will choose the number of addresses to dial taking into account the current number of outgoing connections and ongoing dialings. The addresses to dial will be picked with a bias towards new and vetted peers (TODO define ).
+
+Except for persistent peers, all other peers can be dialed at most `maxAttemptsToDial`. If a node is not connected by that time it is marked as bad. Otherwise, if the dial fails, the node will wait before the next dial. (exponential backoff mechanism). 
+
+If a node needs more peers, the PEX reactor checks first whether peers marked as bad can be reinstated and then also picks a random peer from the store to dial. The PEX reactor relies on the switch to do the actual dialing of a peer.
+
+Once the addresses to dial are known, they are forwarded to the `DialPeersAsync` routine of the switch. Each address is then dialed in parallel and the corresponding peer is added to the `dialing` list. 
+
+**Note** When dialing a peer, the peers' address is not added into the address book as it had to be there in order for the node to dial it in the first place. 
+
+
+#### *Successful dialing*
+
+When a peer is successfully dialed, it is removed from the `dialing` list and added to the `peers` list. The switch then calls the `InitPeer` and `AddPeer` routines of all the reactors that have registered to it. 
+
+Once a peer is successfully dialed, we ask it to provide us with more peers.
+
+#### *Dialing failed*
+
+
+### *Removing peers*
+
+Each reactor can call the `StopPeerForError` method of the switch with the ID of the peer that needs to be removed. Then the switch handles stopping the peer (closing the connection to it), and calls the `RemovePeer` method of all reactors registered to the switch.
```

Comment:
I have to confirm, but the peer instance itself my also trigger this method in case of errors in the connection.

In fact, the `onPeerError` field in the `peerConfig` struct is configured as a pointer to this method. This same pointer is passed to the connection, becoming the `onError` field of the `MConnection` struct. This method is then called in several error situations in the connection managament.